mesonbuild/interpreter.py

```diff
@@ -2460,6 +2456,9 @@ def assignment(self, node):
         value = self.to_native(value)
         if not self.is_assignable(value):
             raise InvalidCode('Tried to assign an invalid value to variable.')
+        # For mutable objects we need to make a copy on assignment
+        if isinstance(value, EnvironmentVariablesHolder) or isinstance(value, ConfigurationDataHolder):
+            value = copy.deepcopy(value)
```

Comment:
I'm wondering if we should add a `mutable` flag on various variable holders and enforce that everywhere.